Lazy Images: Move into a package

davidlonjon opened the pull request:
#### Changes proposed in this Pull Request:
* Moves the Lazy Images module into a package. @gravityrail should be able to provide more context. It has become a requirement for an (yet) undisclosed A8C project.

#### Does this pull request change what data or activity we track or use?
No it should not.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to `wp-admin/admin.php?page=jetpack_modules`
* Activate the "Lazy images" module
* Create a new post with an image inside it.
* Open that post as a non logged in user
* Look at the source of the page and verify that the img tag as the `data-lazy-loaded="1"` property.

#### Proposed changelog entry for your changes:
* Moves the Lazy Images module into a package.

#### Additional development context:

Here are some additional information related to the decision I have made for migrating the `Lazy Images` module into a standalone package. I have made some assumptions which might be incorrect and therefore some changes will need to be amended based on your guidance.

- While migrating the module to a package, I tried to remove or update references to the module source file in order place of the repository such as from:
    - `.eslintignore`
    - `modules/photon-cdn/jetpack-manifest.php`: I am not sure if this one would be generated automatically so i just updated the path to the JavaScript files manually.
   - `phpunit.xml.dist`
   - `modules/lazy-images.php`: I kept this file even though the module source moved into a package as I wanted to make sure the module pattern activation/deactivation still work. I have also moved some of the credit comments into the package itself (`packages/lazy-images/src/lazy-images.php`) as it seems appropriate to belong there as the package can be a standalone download in any project.

- In `packages/lazy-images/src/lazy-images.php`:
    - I have used the version `8.8` in the PHPDoc `@since` entry but I am unsure if this would be the correct version.
    - I have created a new method  called `is_amp_request` because as a standalone package it cannot rely anymore on the `Jetpack_AMP_Support::is_amp_request()` method.
    - I have created a new method  called `get_file_url_for_environment` because as a standalone package it cannot rely anymore on the `Assets::get_file_url_for_environment` method as the JS asset is located within the package.
    - I have fixed the coding standards.

- In `packages/lazy-images/tests/php/test_class.lazy-images.php`:
    - I have fixed the coding standards.

- In `packages/lazy-images/src/js/lazy-images.js`:
    - I have fixed a spelling mistake
    - I have fixed Fix an issue with a comma being in place of a semi-column ( see commit 746e75e7b)

- I have introduced some tooling for building assets, running unit tests and linters. The reason is because I was not sure if/how to integrate these within the Jetpack workflow. So these changes might need to be altered.

    - Previously the minified version of the `lazy-images.js` file was generated by the Jetpack build workflow. Here I have introduced a Webpack config to build the minified version. The `yarn build` command allows building the minified version of the JavaScript.
    - Because the PHP test class (`packages/lazy-images/tests/php/test_class.lazy-images.php`) extends the `WP_UnitTestCase` class, it seems that the `wordpress-tests-lib` is required. Therefore I also needed to have a bootstrapped  WordPress testing environment which I am creating with Docker. The `yarn test:phpunit` command allows to run the PHP unit tests.
    - I have introduced linting the PHP code. The `yarn php:lint` command allows to run the lint the PHP code base.

As I mention some of my assumptions and decisions above might be incorrect and will need to be adapted to an existing Jetpack workflow.

jeherve commented on modules/photon-cdn/jetpack-manifest.php:
```diff
@@ -1,4 +1,4 @@
-<?php 
```
This file is automatically updated as part of the release process, so you can ignore it, there is no need to update it here.

davidlonjon replied:
Thank you for the confirmation. I have now reverted the changes

jeherve commented on packages/lazy-images/.phpcs.xml.dist:
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0"?>
```
Do we need this custom configuration here? Can we rely on the existing configuration instead?

davidlonjon replied:
Looking more into this, I now understand better the processes on Jetpack :) I have now removed it so the existing configuration will be use instead.

jeherve commented on packages/lazy-images/webpack.config.prod.js:
```diff
@@ -0,0 +1,12 @@
+const path = require("path");
```
I wonder if we should leave this out of the package, and handle file minification from the Jetpack plugin instead?

gravityrail replied:
Agreed, let's do that and keep the build scripts out of the package.

davidlonjon replied:
I have created a new webpack config for packages at the root of the repo and added a new command to the package.json. Hopefully this is closer to what it should be. The webpack config for packages can be extended later if required i guess?

davidlonjon commented:
Thanks for the feedback @jeherve.

I have addressed all of your feedback except for the failure of the unit tests. 

Today I tried to use  [WorDBless](https://github.com/Automattic/wordbless) and all the tests would work with it except for `test_wp_get_attachment_image_gets_lazy_treatment` and `test_wp_get_attachment_image_does_not_get_lazy_treatment_when_skip_lazy_added`

These 2 tests are making use of  `wp_get_attachment_image`, `wp_get_attachment_image_srcset` or the gallery shortcode  for which their output are being tested. I am not sure to see how it can be tested by only using `WorDBless` and not the `create_upload_object` to saves an attachment in the DB. I guess we could mock these but it would defeat the purpose. 
The test are still failing with my current setup bootstrapping WordPress into docker but now it it is only a problem with the DB not being ready in time which I think I will be able to fix soon. If you have any suggestions/approach, please let me know.

leogermani commented:
hi @jeher and @davidlonjon 

Nice to hear that the majority of the tests worked with WorDBless out of the box. But the remaining tests will not work indeed. At least not for now. 

I have plans to add support to adding and fetching posts and post metadata, but that will require some work still. (I started on this today to see if it would be quick but need more time. Help is welcome)

It would be great if it was possible to mock the core functions here and leave the environment much simpler and the tests much faster, but I understand that it might not be possible because we are precisely testing filters to core functions, and if we mock them, we are not testing what we want.

davidlonjon commented:
Hi @leogermani, @jeherve , @gravityrail ,

I think I have now reached, I hope, a good compromise for running the PHP Unit tests. What I have ended up doing:
- When running the Lazy Images package PHP unit tests on Travis the following happens:
     - a custom `bin/install-wp-tests.sh` script  installs the WordPress Tests Lib in the root of the Lazy Images package (at wordpress-tests-lib) and sets in `wordpress-tests-lib/wp-tests-config.php` the `ABSPATH` and the DB environment variables to the ones of the WordPress instance running in Travis.
   - Install composer dependencies
  - Run PHPUnit

- When running the Lazy Images package PHP unit tests on a local environment the following happens:
    - a custom `bin/install-wp-tests.sh` script  installs the WordPress Tests Lib in the root of the Lazy Images package (at wordpress-tests-lib) and sets in `wordpress-tests-lib/wp-tests-config.php` the `ABSPATH` and the DB environment variables to the ones of a WordPress environment bootstrapped into Docker.
    - a Docker WordPress environment is being bootstrapped and started using a Docker Compose file local to the Lazy Images Package
    - install composer dependencies
    - run PHPUnit inside the Docker WordPress container
    - stop the Docker container instances

It seems to work both locally and on Travis (See for example the job https://travis-ci.com/github/Automattic/jetpack/jobs/369226564 though at the time of writing this comment, the Travis build failed because of job https://travis-ci.com/github/Automattic/jetpack/jobs/369226569 failing but it seems unrelated to the Lazy Images Package).

Locally I could have maybe re-used the Jepack Docker environment, but I made the assumptions that not all the developer might use it. That's the reason why I decided to have a standalone WordPress environment just for running the PHP unit tests.

I ended up not using `WorDBless` since now a WordPress instance is always available for the Unit tests.

I hope now this approach is acceptable and maybe it could be use as a blueprint for other packages where `WorDBless` might not be enough.

I am looking forward to hearing feedback for the above and also the rest of the PR if there is anything else to address.

I'll re-run the build hoping for job https://travis-ci.com/github/Automattic/jetpack/jobs/369226569  to be fixed.

davidlonjon commented:
Hi @gravityrail ,

Thanks for the additional feedback. I have actually did get rid of the `/bin/wait-for-it.sh` script and replaced by another script much simpler `./bin/wait-for-db.sh` in the push I have done yesterday so hopefully it should solve the issue.

jeherve commented:
Thank you for all the changes!

> I am looking forward to hearing feedback for the above and also the rest of the PR if there is anything else to address.

Looking at the tests, and the extra infrastructure you've had to add to make them work outside of the Jetpack context where they used to run, I think this may be a good time to revisit those tests. Such added testing infrastructure makes for added maintenance, and as the number of packages in Jetpack increases, we should try to avoid adding too much maintenance for each package.

As we move the Lazy Images feature to its own package, thus isolating it more from the Jetpack plugin, I think it would make sense to isolate the tests more as well, to reach a point where we do not need to load WordPress to run all those tests. 

For some of those tests, we may be able to switch and rely on WorDBless as you've investigated, but in general I think it may be fine to test only the data manipulation and remove the WordPress-specific parts that make us need the WordPress environment to test things. 

What do you think?

davidlonjon commented:
Hi @jeherve,

Thanks for the extra feedback and sorry for a late reply.

I am happy with your suggestion and can definitely use `WorDBless` as after i tried it, all the tests work with it except for
 - test_wp_get_attachment_image_gets_lazy_treatment 
 - test_wp_get_attachment_image_does_not_get_lazy_treatment_when_skip_lazy_added

I agree with your the extra infrastructure needed requires more maintenance and I can understand that it is less than ideal so I am happy with your suggestion. If I am understand it well, you are suggesting that we are removing the 2 tests mentioned above (or at least marking them as skipped until we found another way)? 

I am not sure if there would be an alternative for them. I have investigated before mocking the `wp_get_attachment_image ` function for the `test_wp_get_attachment_image_gets_lazy_treatment` test and it seems to work well enough (but defeat the purpose really of testing the filter in that method). For `test_wp_get_attachment_image_does_not_get_lazy_treatment_when_skip_lazy_added` I cannot think of a way to replace the test. Maybe someone who is more expert in that area could chip in?

Alternatively maybe we could skip these 2 tests when running locally but make them run on Travis since we have a WordPress instance running, the tests are passing.

cc @gravityrail

gravityrail commented:
@davidlonjon @jeherve  just disable those two tests for now and let's get this merged.

davidlonjon commented:
Hi @jeherve, @gravityrail,

Based on your advices, I have refactored  the way that PHP unit tests are run, making use of `WorDBless` and disabling the 2 tests which cannot run in that context. I have also removed all the logic to setup a local WordPress instance. (see 
2603c45). So now the process and code is in line with the other packages. 

Additionally I have:
- Fixed instructions in README (4f0ff88)
- Added some ignore rules for phpcs checks (18c6f99). If these ignores are not set, then when doing the PHPCS checks locally after running the unit tests, they also scan the `wordpress` directory created by `WorDBless` and create some errors.
- Added repositories reference to streamline with other packages (045399f): I have noticed this is how it is done in the `connection` package so I brought this forward.
- Fixed PHPCS warnings by adding PHPCS ignore comments and renaming files to appropriate file type (5bfe7aa). Having the `tests/php/post-image-placeholder-content.html` and `tests/php/pre-image-placeholder-content.html` files having a PHP extension but not PHP tags in them created PHPCS warnings and did not seem to need the PHP file extension so I renamed them.

The process for running the unit tests is now streamlined with the other packages and the tests (except the ones skipped) are passing both locally and on Travis.

Hopefully this PR is now good enough for considering merging it.

davidlonjon commented:
Also I am not too sure why some Travis tests are failing. I feel they are unrelated to the changes in this PR but I could be wrong.

jeherve commented on composer.json:
```diff
@@ -41,7 +42,7 @@
 	},
 	"scripts": {
-		"php:compatibility": "vendor/bin/phpcs -p -s --runtime-set testVersion '5.6-' --standard=PHPCompatibilityWP --ignore=docker,tools,tests,node_modules,vendor --extensions=php",
+		"php:compatibility": "vendor/bin/phpcs -p -s --runtime-set testVersion '5.6-' --standard=PHPCompatibilityWP --ignore=docker,tools,tests,node_modules,vendor,packages/*/wordpress --extensions=php",
```
You should be able to get rid of this change now.

davidlonjon replied:
If I remove this and the `wordpress` directory created by `WorDBless` is at the root of this package, then this command fails because it scans the `wordpress` directory. I can off course remove this if it makes sense for the process on Travis.

jeherve commented on package.json:
```diff
@@ -24,8 +24,9 @@
-		"build": "yarn install-if-deps-outdated && yarn clean && yarn build-client && yarn build-php && yarn build-extensions && yarn build-search",
+		"build": "yarn install-if-deps-outdated && yarn clean && yarn build-client && yarn build-php && yarn build-extensions && yarn build-search && yarn build-packages",
 		"build-client": "gulp",
+		"build-packages": "webpack --config ./webpack.config.packages.js",
```
We may have to add this to the `build-production` command as well.

davidlonjon replied:
I completely missed that one. I have now added. Thank you

jeherve commented on packages/lazy-images/.gitattributes:
```diff
@@ -0,0 +1,7 @@
+tests/         export-ignore
+composer.lock  export-ignore
+vendor/        export-ignore
+wordpress/    export-ignore
```
You may not need this anymore?

davidlonjon replied:
I actually mis-undertood the meaning of this. I understand now. Thanks for pointing it out

jeherve commented on packages/lazy-images/.gitattributes:
```diff
@@ -0,0 +1,7 @@
+# Files not needed to be distributed in the package.
+.gitattributes export-ignore
+phpunit.xml    export-ignore
```
Should it be `phpunit.xml.dist`?

davidlonjon replied:
Yes you are right. Actually my mistake was due to copy the original .gitattributes file from the `connection` package which has this mistake and might need fixing too. I have now fixed that for this package.

jeherve replied:
> Actually my mistake was due to copy the original .gitattributes file from the connection package which has this mistake and might need fixing too. 

Yes, that's a recent oversight:
https://github.com/Automattic/jetpack/pull/16810#issuecomment-673931819

jeherve commented on packages/lazy-images/.gitattributes:
```diff
@@ -0,0 +1,7 @@
+.gitattributes export-ignore
+phpunit.xml    export-ignore
+tests/         export-ignore
+composer.lock  export-ignore
+vendor/        export-ignore
```
Do we need those 2 since we do not check them in?

davidlonjon replied:
Yes you are right. I have now removed these entries.

leogermani commented:
Hi everyone,

I have been able to make some progress in WorDBless this week. There's a chance I can make it able to process the missing tests. If we have time and can wait a couple of days I might be able to bring some good news :)

leogermani commented:
@davidlonjon @gravityrail I have some good news.

In https://github.com/Automattic/jetpack/pull/16843 you can see one of the tests passing. WorDBless now supports posts, postmeta and attachments. (this is so much fun)

The second test will not pass because the gallery shortcode relies on get_children(), which uses get_posts() and WP_Query. WorDBless doesn't support it and, although I can see a way we could partially add support to it (querying posts by ID, post types and other simple queries) this will take a long time...

Also, the gallery shortcode actually uses the same wp_get_attachment_image function, so the test seems to be testing the same thing (if I got it correctly)

If you could have a look, I will wrap this up and merge the latest changes to WorDBless master and then I think we are good to go!

gravityrail commented:
@leogermani I see a bunch of failures on Travis like this:

```
Could not scan for classes inside "/home/travis/build/Automattic/jetpack/vendor/automattic/jetpack-lazy-images/src/" which does not appear to be a file nor a folder
```

Can you please take a look at that? Thanks.

leogermani commented:
> @leogermani I see a bunch of failures on Travis like this:
> 
> ```
> Could not scan for classes inside "/home/travis/build/Automattic/jetpack/vendor/automattic/jetpack-lazy-images/src/" which does not appear to be a file nor a folder
> ```
> 
> Can you please take a look at that? Thanks.

I see a couple of different errors that seems unrelated with the recent changes and some comments that they were there before.

Maybe it's a good idea rebasing it against master

gravityrail commented:
Updated from master

gravityrail commented:
@davidlonjon @leogermani I am still getting test failures after merging master:

```
Travis CI command: phpunit
Running PHP:Compatibility checks:
PHP Compatibility command: `composer php:compatibility .` 
PHP_CodeSniffer version 3.5.5 (stable) by Squiz (http://www.squiz.net)
Could not scan for classes inside "/home/travis/build/Automattic/jetpack/vendor/automattic/jetpack-lazy-images/src/" which does not appear to be a file nor a folder
> vendor/bin/phpcs -p -s --runtime-set testVersion '5.6-' --standard=PHPCompatibilityWP --ignore=docker,tools,tests,node_modules,vendor,packages/*/wordpress --extensions=php '.'
ERROR: the "PHPCompatibilityWP" coding standard is not installed. The installed coding standards are PSR12, PEAR, Zend, PSR1, MySource, PSR2 and Squiz
Script vendor/bin/phpcs -p -s --runtime-set testVersion '5.6-' --standard=PHPCompatibilityWP --ignore=docker,tools,tests,node_modules,vendor,packages/*/wordpress --extensions=php handling the php:compatibility event returned with error code 3
The command "./tests/run-travis.sh" exited with 1.
```

gravityrail commented:
Actually there are even more different issues, if you click around the test failures. At least a few seem unrelated to these changes and should be fixed in master.

davidlonjon commented:
First, thank you so much @leogermani for the work you have done with `WorDBless` and as a result refactoring one of the tests here. This is awesome! 

Regarding the `test_wp_get_attachment_image_does_not_get_lazy_treatment_when_skip_lazy_added`, I think you are right and the test is really similar from `test_wp_get_attachment_image_gets_lazy_treatment`. I kept them those 2 tests as they were already there. 

But your comment actually prompted me to look again at it and since we only want to test the `wp_get_attachment_image` function output in the context that it does not get the get the lazy treatment when lazy images feature is skipped using the `skip-lazy` class, then I remove the use of the gallery shortcode and just tested the `wp_get_attachment_image` function directly. It seems to work so now there is no dependency on the gallery shortcode and the calls to `wp_posts`.

@gravityrail, regarding the tests failure you've seen. It turns out the reason was that the `composer.lock` was out of date. It took me a long while to figure it out by testing all the Travis build process locally double checking on different branches. 

Finally all the tests are now passing again and it seems the integration works well enough again. Hopefully this is now it and we should be in a state that it could be merged?

cc @jeherve

leogermani commented:
Seems ready to go. @jeherve just confirm your request change was addressed.

davidlonjon commented:
I pushed a commit merging master back into this branch to fix a merge conflict with the `package-lock.json` file. All seems good and the Travis CI build is passing again

jeherve commented:
Internal reference: D48486-code
